src/util/codec/number.rs

```diff
@@ -219,6 +222,128 @@ pub trait NumberDecoder: Read {
 
 impl<T: Read> NumberDecoder for T {}
 
+type Bytes<'a> = &'a [u8];
+
+fn read_num_bytes<T, F>(size: usize, data: &mut &[u8], f: F) -> Result<T>
+where
+    F: Fn(&[u8]) -> T,
+{
+    if data.len() >= size {
+        let buf = &data[0..size];
+        *data = &data[size..];
+        return Ok(f(buf));
+    }
+    Err(Error::Io(io::Error::new(ErrorKind::UnexpectedEof, "eof")))
+}
+
+/// `decode_i64` decodes value encoded by `encode_i64` before.
+#[inline]
+fn decode_i64(data: &mut Bytes) -> Result<i64> {
+    decode_u64(data).map(order_decode_i64)
+}
+
+/// `decode_i64_desc` decodes value encoded by `encode_i64_desc` before.
+#[inline]
+fn decode_i64_desc(data: &mut Bytes) -> Result<i64> {
+    decode_u64_desc(data).map(order_decode_i64)
+}
+
+/// `decode_u64` decodes value encoded by `encode_u64` before.
+#[inline]
+pub fn decode_u64(data: &mut Bytes) -> Result<u64> {
+    read_num_bytes(mem::size_of::<u64>(), data, BigEndian::read_u64)
+}
+
+/// `decode_u64_desc` decodes value encoded by `encode_u64_desc` before.
+#[inline]
+pub fn decode_u64_desc(data: &mut Bytes) -> Result<u64> {
+    let v = decode_u64(data)?;
+    Ok(!v)
+}
+
+/// `decode_var_i64` decodes value encoded by `encode_var_i64` before.
+#[inline]
+fn decode_var_i64(data: &mut Bytes) -> Result<i64> {
+    let v = decode_var_u64(data)?;
+    let mut vx = v >> 1;
+    if v & 1 != 0 {
+        vx = !vx;
+    }
+    Ok(vx as i64)
+}
+
+/// `decode_var_u64` decodes value encoded by `encode_var_u64` before.
+#[inline]
+pub fn decode_var_u64(data: &mut Bytes) -> Result<u64> {
+    if data.len() >= 1 && data[0] < 0x80 {
+        let res = u64::from(data[0]);
+        *data = &data[1..];
+        return Ok(res);
+    }
+
+    let (mut x, mut s, mut i) = (0, 0, 0);
+    while i < data.len() {
```

Comment:
If `data.len() >= 10` or last byte is less than 0x80, you can use a static loop `for i in 0..9` and check the final byte instead.